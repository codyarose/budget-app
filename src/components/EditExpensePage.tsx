import React, { ReactElement } from 'react'

const EditExpensePage = ({
	match,
}: {
	match: {
		params: {
			id: string
		}
	}
}): ReactElement => {
	return <div>Editing expense with id of {match.params.id}</div>
}

export default EditExpensePage
