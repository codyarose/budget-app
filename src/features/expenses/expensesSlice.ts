import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { v4 as uuid } from 'uuid'
import { Moment } from 'moment'

import { RootState } from '../../app/store'
import moment from 'moment'

export interface Expense {
	id: string
	description: string
	note: string
	amount: number
	createdAt: Moment | number
}

export type ExpenseData = Omit<Expense, 'id'>

export const initialState: Expense[] = []

export const expensesSlice = createSlice({
	name: 'expenses',
	initialState,
	reducers: {
		addExpense: (state, action: PayloadAction<Partial<ExpenseData>>) => {
			const {
				description = '',
				note = '',
				amount = 0,
				createdAt = 0,
			} = action.payload
			state.push({
				id: uuid(),
				description,
				note,
				amount,
				createdAt,
			})
		},
		removeExpense: (state, action: PayloadAction<string>) => {
			const id = action.payload
			const index = state.findIndex((expense) => expense.id === id)
			if (index === -1) {
				return state
			} else {
				state.splice(index, 1)
			}
		},
		editExpense: (
			state,
			action: PayloadAction<{
				editedExpense: Partial<ExpenseData>
				id: string
			}>,
		) => {
			const { editedExpense, id } = action.payload
			return state.map((expense) => {
				if (expense.id === id) {
					return {
						...expense,
						...editedExpense,
					}
				} else {
					return expense
				}
			})
		},
	},
})

export const { addExpense, removeExpense, editExpense } = expensesSlice.actions

export const selectExpenses = (state: RootState): Expense[] => state.expenses

export const selectVisibleExpenses = (state: RootState): Expense[] => {
	const expenses = state.expenses
	const { startDate, endDate, text, sortBy } = state.filter

	return expenses
		.filter((expense) => {
			const createdAtMoment = moment(expense.createdAt)
			const momentStartDate = moment(startDate)
			const momentEndDate = moment(endDate)

			const startDateMatch = startDate
				? momentStartDate.isSameOrBefore(createdAtMoment, 'day')
				: true
			const endDateMatch = endDate
				? momentEndDate.isSameOrAfter(createdAtMoment, 'day')
				: true
			const textMatch = expense.description
				.toLowerCase()
				.includes(text.toLowerCase())

			return startDateMatch && endDateMatch && textMatch
		})
		.sort((a, b) => {
			if (sortBy === 'date') {
				return a.createdAt < b.createdAt ? 1 : -1
			} else if (sortBy === 'amount') {
				return a.amount < b.amount ? 1 : -1
			} else {
				return 1
			}
		})
}

export const selectTotal = (state: RootState): number => {
	const expenses = selectVisibleExpenses(state)

	if (!expenses.length) return 0

	return expenses.reduce((acc, expense) => {
		return acc + expense.amount
	}, 0)
}

export default expensesSlice.reducer
