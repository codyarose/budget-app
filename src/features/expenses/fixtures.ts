export default [
	{
		id: '123abc',
		description: 'Title',
		note: '',
		amount: 140500,
		createdAt: 1598763926001,
	},
	{
		id: '456def',
		description: 'Blah',
		note: 'This is a note',
		amount: 2650,
		createdAt: 1598763926009,
	},
	{
		id: '789ghi',
		description: 'Another one',
		note: 'Text goes here',
		amount: 340098,
		createdAt: 1598763926345,
	},
]
