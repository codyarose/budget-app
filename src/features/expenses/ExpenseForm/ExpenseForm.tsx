import React, { FC, useState, ChangeEvent, FormEvent } from 'react'
import styled from 'styled-components'
import moment, { Moment } from 'moment'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/initialize'

import { ExpenseData } from '../expensesSlice'
import Error from '../../../components/Error'

interface Props {
	expense?: ExpenseData
	onSubmit: (arg0: ExpenseData) => void
}

interface FormState extends Omit<ExpenseData, 'amount'> {
	amount: string
}

const ExpenseForm: FC<Props> = ({ expense, onSubmit }) => {
	const [formState, setFormState] = useState<FormState>({
		description: expense?.description || '',
		amount: expense ? (expense?.amount / 100).toString() : '',
		note: expense?.note || '',
		createdAt: expense?.createdAt || moment().valueOf(),
	})

	const [error, setError] = useState('')

	const [calendarFocused, setCalendarFocused] = useState<boolean | null>(
		false,
	)

	const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
		const { name, value } = e.target
		setFormState({
			...formState,
			[name]: value,
		})
	}

	const handleAmountChange = (e: ChangeEvent<HTMLInputElement>) => {
		const amount = e.target.value
		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			setFormState({ ...formState, amount })
		}
	}

	const handleSubmit = (e: FormEvent) => {
		e.preventDefault()
		const { description, amount } = formState
		if (!description || !amount) {
			setError('Please provide description and amount.')
		} else {
			setError('')
			const parsedAmount = parseFloat(amount) * 100
			onSubmit({
				...formState,
				amount: parsedAmount,
			})
		}
	}

	const handleDateChange = (createdAt: Moment | null) => {
		createdAt &&
			setFormState({ ...formState, createdAt: createdAt.valueOf() })
	}

	const handleFocusChange = (arg: { focused: boolean | null }) => {
		setCalendarFocused(arg.focused)
	}

	return (
		<div>
			<StyledForm onSubmit={handleSubmit}>
				<input
					type="text"
					placeholder="Description"
					name="description"
					id="expenseDescription"
					autoFocus
					value={formState.description}
					onChange={handleInputChange}
				/>
				<input
					type="text"
					placeholder="Amount"
					name="amount"
					id="expenseAmount"
					value={formState.amount || ''}
					onChange={handleAmountChange}
				/>
				<Error error={!!error} helperText={error} />
				<SingleDatePicker
					date={moment(formState.createdAt)}
					onDateChange={handleDateChange}
					focused={calendarFocused}
					onFocusChange={handleFocusChange}
					id="datePicker"
					numberOfMonths={1}
					isOutsideRange={() => false}
				/>
				<input
					type="text"
					name="note"
					id="expenseNote"
					placeholder="Add a not for your expense (optional)"
					value={formState.note}
					onChange={handleInputChange}
				/>
				<button type="submit">Add expense</button>
			</StyledForm>
		</div>
	)
}

export default ExpenseForm

const StyledForm = styled.form`
	display: flex;
	flex-flow: column;
`

StyledForm.displayName = 'StyledForm'
