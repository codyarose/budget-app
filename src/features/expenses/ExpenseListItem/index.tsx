import React, { FC } from 'react'
import { Link } from 'react-router-dom'

import { Expense } from '../expensesSlice'

const ExpenseListItem: FC<Expense> = ({
	description,
	amount,
	createdAt,
	note,
	id,
}) => {
	return (
		<div>
			<Link to={`/edit/${id}`}>
				<h4>{description}</h4>
			</Link>
			<p>
				{amount} - {createdAt}
			</p>
			{note && <p>{note}</p>}
		</div>
	)
}

export default ExpenseListItem
