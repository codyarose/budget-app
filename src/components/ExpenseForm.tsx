import React, { FC, useState, ChangeEvent, FormEvent } from 'react'
import styled from 'styled-components'
import moment, { Moment } from 'moment'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'
import 'react-dates/initialize'

import { Expense } from '../features/expenses/expensesSlice'
import Error from './Error'

interface Props {
	expense?: Expense
	onSubmit: (arg0: Partial<Expense>) => void
}

interface FormState extends Omit<Expense, 'amount'> {
	amount: string
}

const ExpenseForm: FC<Props> = ({ expense, onSubmit }) => {
	const [formState, setFormState] = useState<Partial<FormState>>({
		description: expense?.description || '',
		amount: expense ? (expense?.amount / 100).toString() : '',
		note: expense?.note || '',
		createdAt: expense?.createdAt || moment().valueOf(),
	})

	const [error, setError] = useState('')

	const [calendarFocused, setCalendarFocused] = useState<boolean | null>(
		false,
	)

	const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
		const { name, value } = e.target
		setFormState({
			...formState,
			[name]: value,
		})
	}

	const handleAmountChange = (e: ChangeEvent<HTMLInputElement>) => {
		const amount = e.target.value
		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			setFormState({ ...formState, amount })
		}
	}

	const handleSubmit = (e: FormEvent) => {
		e.preventDefault()
		const { description, amount } = formState
		if (!description || !amount) {
			setError('Please provide description and amount.')
		} else {
			setError('')
			const parsedAmount = parseFloat(amount) * 100
			onSubmit({
				...formState,
				amount: parsedAmount,
			})
		}
	}

	const handleDateChange = (createdAt: Moment | null) => {
		createdAt &&
			setFormState({ ...formState, createdAt: createdAt.valueOf() })
	}

	const handleFocusChange = (arg: { focused: boolean | null }) => {
		setCalendarFocused(arg.focused)
	}

	return (
		<div>
			<StyledForm onSubmit={handleSubmit}>
				<input
					type="text"
					placeholder="Description"
					name="description"
					autoFocus
					value={formState.description}
					onChange={handleInputChange}
				/>
				<input
					type="text"
					placeholder="Amount"
					name="amount"
					value={formState.amount || ''}
					onChange={handleAmountChange}
				/>
				<Error error={!!error} helperText={error} />
				<SingleDatePicker
					date={moment(formState.createdAt)}
					onDateChange={handleDateChange}
					focused={calendarFocused}
					onFocusChange={handleFocusChange}
					id="datePicker"
					numberOfMonths={1}
					isOutsideRange={() => false}
				/>
				<input
					type="text"
					name="note"
					placeholder="Add a not for your expense (optional)"
					value={formState.note}
					onChange={handleInputChange}
				/>
				<button type="submit">Add expense</button>
			</StyledForm>
		</div>
	)
}

export default ExpenseForm

const StyledForm = styled.form`
	display: flex;
	flex-flow: column;
`
