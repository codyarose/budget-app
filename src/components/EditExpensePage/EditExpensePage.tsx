import React, { ReactElement, FC } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import styled from 'styled-components'
import { RouteComponentProps } from 'react-router-dom'
import { lighten } from 'polished'

import {
	selectExpenses,
	editExpense,
	removeExpense,
	ExpenseData,
} from '../../features/expenses/expensesSlice'
import ExpenseForm from '../../features/expenses/ExpenseForm'
import ContentContainer from '../common/ContentContainer'
import Button from '../common/Button'
import PageHeader from '../common/PageHeader'

interface Props {
	match: {
		params: {
			id: string
		}
	}
}

const EditExpensePage: FC<Props & RouteComponentProps> = ({
	match,
	history,
}): ReactElement => {
	const expenses = useSelector(selectExpenses)
	const dispatch = useDispatch()

	const id = match.params.id

	const expenseToEdit = expenses.find((expense) => expense.id === id)

	const handleSubmit = (editedExpense: ExpenseData) => {
		dispatch(editExpense({ editedExpense, id }))
		history.push('/')
	}

	const handleRemove = () => {
		dispatch(removeExpense(id))
		history.push('/')
	}

	return (
		<ContentContainer padBottom={true}>
			<PageHeader>
				<h2>Edit Expense</h2>
			</PageHeader>
			<ExpenseForm expense={expenseToEdit} onSubmit={handleSubmit} />
			<StyledRemoveButton onClick={handleRemove}>
				Remove
			</StyledRemoveButton>
		</ContentContainer>
	)
}

export default EditExpensePage

const StyledRemoveButton = styled(Button)`
	display: block;
	width: 100%;
	max-width: 40%;
	min-width: 180px;
	color: ${({ theme }) =>
		theme.colors.bg === '#000' ? lighten(0.4, 'darkred') : 'darkred'};
	padding: ${({ theme }) => `${theme.spacing.sm}`};
	margin: ${({ theme }) => `${theme.spacing.sm} auto 0`};
`
StyledRemoveButton.displayName = 'RemoveButton'
