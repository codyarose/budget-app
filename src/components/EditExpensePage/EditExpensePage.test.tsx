import React from 'react'
import { shallow } from 'enzyme'
import * as redux from 'react-redux'

import EditExpensePage from './index'
import {
	editExpense,
	removeExpense,
} from '../../features/expenses/expensesSlice'
import expenses from '../../features/expenses/fixtures'
import ExpenseForm from '../../features/expenses/ExpenseForm'

jest.mock('react-redux', () => ({
	useDispatch: jest.fn(),
	useSelector: jest.fn().mockImplementation((selector) => selector()),
}))
jest.mock('../../features/expenses/expensesSlice', () => {
	return {
		selectExpenses: jest.fn().mockReturnValue([
			{
				id: '123abc',
				description: 'Title',
				note: '',
				amount: 140500,
				createdAt: 1598763926001,
			},
		]),
		editExpense: jest.fn(),
		removeExpense: jest.fn(),
	}
})

describe('EditExpensePage', () => {
	const useDispatchSpy = jest.spyOn(redux, 'useDispatch')
	const mockDispatchFn = jest.fn()
	useDispatchSpy.mockReturnValue(mockDispatchFn)

	let props: any, wrapper: any
	beforeEach(() => {
		props = {
			match: {
				params: {
					id: expenses[0].id,
				},
			},
			onSubmit: mockDispatchFn,
			history: { push: jest.fn() },
		}
		wrapper = shallow(<EditExpensePage {...props} />)
	})

	it('should render correctly', () => {
		expect(wrapper).toMatchSnapshot()
	})

	it('should handle editExpense', () => {
		wrapper.find(ExpenseForm).prop('onSubmit')(expenses[0])

		expect(props.history.push).toHaveBeenLastCalledWith('/')
		expect(editExpense).toHaveBeenLastCalledWith({
			editedExpense: expenses[0],
			id: expenses[0].id,
		})
	})

	it('should handle removeExpense', () => {
		wrapper.find('button').simulate('click')

		expect(props.history.push).toHaveBeenLastCalledWith('/')
		expect(removeExpense).toHaveBeenLastCalledWith(expenses[0].id)
	})
})
