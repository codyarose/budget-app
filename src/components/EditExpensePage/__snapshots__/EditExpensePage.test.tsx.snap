// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditExpensePage should render correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 140500,
        "createdAt": 1598763926001,
        "description": "Title",
        "id": "123abc",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
